### PR TITLE
[WIP] Optimize after generating AutoDiff code from Symbolic script

### PR DESCRIPTION
**Pull request by t-vi**
Our symbolic script is large enough to deserve optimization after it is applied. In particular, obsolete if branches from constant propagation seem to be a thing.

The grad_sum_to_size elimination (#18697) will benefit from the implicit specialization.

### CODE REVIEW

**Review by eellison — DISMISSED** [review]
LGTM

File: torch/csrc/jit/graph_executor.cpp [hunk 1]
```diff
@@ -608,6 +608,13 @@ struct GraphExecutorImpl {
       for (Node* dnode : diff_nodes) {
         auto diff_graph = std::move(dnode->g(attr::Subgraph));
         Gradient gradient = differentiate(diff_graph);
+
+        // Re-run ConstantPropagation and Optimization with gradient.f
+        // In particular, we want to re-specialize symbolic_script
+        // Optionals relating to _grad_sum_to_size for the graph_fuser
+        ConstantPropagation(gradient.f);
```
Comment (eellison) [inline torch/csrc/jit/graph_executor.cpp]: Any reason to not run propagate input shapes?

Comment (eellison) [reply torch/csrc/jit/graph_executor.cpp]: ?

Comment (t-vi) [reply torch/csrc/jit/graph_executor.cpp]: For some reason I have not fully understood yet, it seems to inhibit the graph fuser from fusing e.g. the mul(mask, *), div(, 1-prob). I didn't have a chance to investigate yesterday.
As this is closely related to #19988, I'll have to rebase anyway and update this.

**Comment by t-vi** [comment]
So this is now a test / comment amendment to #19988.

**Comment by ezyang** [comment]
This error looks real:

```
21:07:35 ======================================================================
21:07:35 FAIL: test_nn_dropout (__main__.TestJitGeneratedFunctional)
21:07:35 ----------------------------------------------------------------------
21:07:35 Traceback (most recent call last):
21:07:35   File "test_jit.py", line 14280, in wrapper
21:07:35     return fn(*args, **kwargs)
21:07:35   File "test_jit.py", line 255, in wrapper
21:07:35     return fn(*args, **kwargs)
21:07:35   File "test_jit.py", line 14331, in do_test
21:07:35     run_test()
21:07:35   File "test_jit.py", line 14325, in run_test
21:07:35     self.assertAutodiffNode(script_fn.last_graph, should_autodiff_node, autodiff_nodes, fusible_nodes)
21:07:35   File "test_jit.py", line 508, in assertAutodiffNode
21:07:35     self.assertEqual(should_autodiff_node, found_all_nonfusible_nodes and found_all_fusible_nodes)
21:07:35   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\common_utils.py", line 500, in assertEqual
21:07:35     super(TestCase, self).assertEqual(x, y, message)
21:07:35 AssertionError: True != False : 
```

**Comment by t-vi** [comment]
Is this related to windows not having fusion? I could make the expectation to depend on not running on Windows, but it's kind of ugly.

**Comment by ezyang** [comment]
Yeah I dunno.

**Comment by ezyang** [comment]
Any update on this PR?

**Comment by t-vi** [comment]
No, I need to check whether updating or dropping is the best way.

**Comment by t-vi** [comment]
I think we'll drop this. It's hairy with Windows and the value is very limited.

**Pull request closed without merging**